incubator/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/config/MqttKafkaHeaderHelper.java

```diff
@@ -0,0 +1,169 @@
+import io.aklivity.zilla.runtime.binding.mqtt.kafka.internal.types.KafkaHeaderFW;
+import io.aklivity.zilla.runtime.binding.mqtt.kafka.internal.types.OctetsFW;
+import io.aklivity.zilla.runtime.binding.mqtt.kafka.internal.types.String16FW;
+import io.aklivity.zilla.runtime.binding.mqtt.kafka.internal.types.stream.KafkaMergedDataExFW;
+
+public class MqttKafkaHeaderHelper
+{
+    public static final String KAFKA_TOPIC_HEADER_NAME = "zilla:topic";
+    public static final String KAFKA_LOCAL_HEADER_NAME = "zilla:local";
+    public static final String KAFKA_TIMEOUT_HEADER_NAME = "zilla:timeout-ms";
+    public static final String KAFKA_CONTENT_TYPE_HEADER_NAME = "zilla:content-type";
+    public static final String KAFKA_FORMAT_HEADER_NAME = "zilla:format";
+    public static final String KAFKA_REPLY_TO_HEADER_NAME = "zilla:reply-to";
+    public static final String KAFKA_CORRELATION_ID_HEADER_NAME = "zilla:correlation-id";
+    public final OctetsFW kafkaTopicHeaderOctets;
+    public final OctetsFW kafkaLocalHeaderOctets;
+    public final OctetsFW kafkaTimeoutHeaderOctets;
+    public final OctetsFW kafkaContentTypeHeaderOctets;
+    public final OctetsFW kafkaFormatHeaderOctets;
+    public final OctetsFW kafkaReplyToHeaderOctets;
+    public final OctetsFW kafkaCorrelationHeaderOctets;
+    private final Map<String, Consumer<OctetsFW>> visitors;
+    private final MutableDirectBuffer octetsBuffer;
+    private final OctetsFW emptyRO = new OctetsFW().wrap(new UnsafeBuffer(0L, 0), 0, 0);
+
+
+    public MqttKafkaHeaderHelper()
+    {
+        octetsBuffer = new UnsafeBuffer(new byte[8 * 1024]);
+        visitors = new HashMap<>();
+        visitors.put(KAFKA_TOPIC_HEADER_NAME, this::visitTopic);
+        visitors.put(KAFKA_LOCAL_HEADER_NAME, this::visitLocal);
+        visitors.put(KAFKA_TIMEOUT_HEADER_NAME, this::visitTimeout);
+        visitors.put(KAFKA_CONTENT_TYPE_HEADER_NAME, this::visitContentType);
+        visitors.put(KAFKA_FORMAT_HEADER_NAME, this::visitFormat);
+        visitors.put(KAFKA_REPLY_TO_HEADER_NAME, this::visitReplyTo);
+        visitors.put(KAFKA_CORRELATION_ID_HEADER_NAME, this::visitCorrelationId);
+
+        kafkaTopicHeaderOctets = stringToOctets(KAFKA_TOPIC_HEADER_NAME);
+        kafkaLocalHeaderOctets = stringToOctets(KAFKA_LOCAL_HEADER_NAME);
+        kafkaTimeoutHeaderOctets = stringToOctets(KAFKA_TIMEOUT_HEADER_NAME);
+        kafkaContentTypeHeaderOctets = stringToOctets(KAFKA_CONTENT_TYPE_HEADER_NAME);
+        kafkaFormatHeaderOctets = stringToOctets(KAFKA_FORMAT_HEADER_NAME);
+        kafkaReplyToHeaderOctets = stringToOctets(KAFKA_REPLY_TO_HEADER_NAME);
+        kafkaCorrelationHeaderOctets = stringToOctets(KAFKA_CORRELATION_ID_HEADER_NAME);
+    }
+
+    public List<String> topicNames;
+    public String local;
+    public int timeout;
+    public String contentType;
+    public String format;
+    public String replyTo;
+    public OctetsFW correlation;
+    public Map<String, List<String>> userProperties;
```

Comment:
Retaining a map of user properties forces us to build up a copy of all of them in memory just to batch them so they can be written out together in the mqtt publish packet.

We are also storing them as type String which requires UTF-8 decoding and encoding on every message passing through the data path.

Instead we can store these as an `IntArrayList` of offsets, then wrap them with a reusable flyweight when we need to get the header values out later, using the offsets to provide random access to the buffer for each user property kafka header.